pipeline/transform.py

```diff
@@ -69,74 +71,120 @@ def __init__(self) -> None:
-            f"*.feather*") if os.path.isfile(file)]
+            "*.feather*") if os.path.isfile(file)]
         data = {}
         for file in files:
             df = pd.read_feather(file)
             if "generation" in file:
                 formatted_data = self.generation_transform(df)
                 data['generation'] = formatted_data
                 self.logger.info(
-                    f"""Successfully transformed generation data""")
+                    """Transformed generation data""")
             if "demand" in file:
                 formatted_data = self.demand_transform(df)
                 data['demand'] = formatted_data
-                self.logger.info(f"""Successfully transformed demand data""")
+                self.logger.info("""Transformed demand data""")
             if "cost" in file:
                 formatted_data = self.cost_transform(df)
-                data['cost'] = formatted_data
-                self.logger.info(f"""Successfully transformed cost data""")
+                if formatted_data:
+                    data['cost'] = formatted_data
+                self.logger.info("""Transformed cost data""")
             if "carbon" in file:
                 formatted_data = self.carbon_transform(df)
                 data['carbon'] = formatted_data
-                self.logger.info(f"""Successfully transformed carbon data""")
+                self.logger.info("""Transformed carbon data""")
+        diff = list(set(self.period_g) - set(self.period_c))
+        for period in diff:
+            for values in data['generation']:
+
+                if values[5] == period:
+                    if period == 2:
+                        yesterday = datetime.date.today() - datetime.timedelta(days=1)
+                        data['cost'].append((yesterday, 2, 0, 0))
+                    data['cost'].append((values[1], values[5], 0, 0))
+                    break
+
+        diff = list(set(self.time_g) - set(self.time_d))
+        for time in diff:
+            for values in data['generation']:
+                if values[0] == time:
+                    data['demand'].append((values[0], 0))
+                    break
 
         self.delete_read_files(files)
         return data
 
     def generation_transform(self, df: pd.DataFrame) -> tuple:
+        """
+        Filters and transforms the generation dataframe passed into it and returns
+        a list of tuples
+        """
         df['publish_date'] = df['publishTime'].apply(lambda x: x.split('T')[0])
         df['gain_loss'] = df['generation'].apply(
             lambda x: '+' if x > 0 else '-')
 
-        # cutoff = datetime.datetime.now() - datetime.timedelta(hours=1, minutes=5)
-        # df = df.query('publishTime != @cutoff')
-        df = df[['publishTime', 'publish_date', 'fuelType', 'gain_loss',
-                'generation', 'settlementPeriod']]
+        self.period_g = df['settlementPeriod'].unique()
+        self.time_g = df['publishTime'].unique()
```

Comment:
I think in an ideal world we might remove magic strings. I myself have used them however. That's OoS for this project.